File: internal/registry/runner/agent.go
package runner

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

// CLIAgentRunner invokes the user's own `claude` or `codex` binary as a one-shot,
// on the user's own subscription/credentials. It mirrors the argv shape used by
// the ZeroShot desktop app's run_agent_oneshot.
type CLIAgentRunner struct {
	Provider string // "claude" | "codex"
	Binary   string // resolved binary path/name
	Version  string // captured agent version
}

// AgentName implements AgentRunner.
func (r *CLIAgentRunner) AgentName() string {
	if r.Provider == "claude" {
		return "claude_code"
	}
	return r.Provider
}

// AgentVersionBand implements AgentRunner. The band collapses patch versions so a
// weekly agent release does not fragment every environment cell (minor-version
// band, mirroring the registry reference pin).
func (r *CLIAgentRunner) AgentVersionBand() string {
	return VersionBand(r.Version)
}

// VersionBand reduces a semver-ish version string to "major.minor".
func VersionBand(version string) string {
	v := strings.TrimSpace(version)
	v = strings.TrimPrefix(v, "v")
	parts := strings.SplitN(v, ".", 3)
	if len(parts) >= 2 {
		return parts[0] + "." + parts[1]
	}
	return v
}

// RunTask installs the skill for the treatment arm, invokes the agent one-shot,
// and parses usage. The verifier is applied by the caller AFTER this returns, so
// the agent workspace never contains verifier content.
func (r *CLIAgentRunner) RunTask(ctx context.Context, spec RunSpec) (RunOutcome, error) {
	if spec.SkillInstalled {
		if err := installSkillIntoWorkspace(spec.WorkingDir, spec.SkillSlug, spec.SkillContent); err != nil {
			return RunOutcome{}, err
		}
	}
	prompt := buildPrompt(spec)

	timeout := time.Duration(spec.TimeoutSec) * time.Second
	if timeout <= 0 {
		timeout = 240 * time.Second
	}
	runCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	args := r.argv(spec)
	cmd := exec.CommandContext(runCtx, r.Binary, args...)
	cmd.Dir = spec.WorkingDir
	cmd.Stdin = strings.NewReader(prompt)

	start := time.Now()
	stdout, err := cmd.Output()
	elapsed := time.Since(start).Milliseconds()

	digest := sha256.Sum256(stdout)
	outcome := RunOutcome{
		DurationMS:       elapsed,
		TrajectoryDigest: "sha256:" + hex.EncodeToString(digest[:]),
	}
	if runCtx.Err() == context.DeadlineExceeded {
		outcome.AgentError = true
		return outcome, nil
	}
	if err != nil {
		// A non-zero exit is treated as an agent error for this trial rather than
		// failing the whole run — completed trials still produce a valid bundle.
		outcome.AgentError = true
		return outcome, nil
	}

	parsed, parseErr := ParseAgentUsage(r.Provider, stdout)
	if parseErr != nil {
		outcome.AgentError = true
		return outcome, nil
	}
	parsed.DurationMS = elapsed
	parsed.TrajectoryDigest = outcome.TrajectoryDigest
	return parsed, nil
}

func (r *CLIAgentRunner) argv(spec RunSpec) []string {
	if r.Provider == "claude" {
		args := []string{"-p", "--output-format", "json", "--dangerously-skip-permissions"}
		if spec.Model != "" {
			args = append(args, "--model", spec.Model)
		}
		return args
	}
	// codex
	args := []string{
		"exec", "--json", "--skip-git-repo-check", "--ignore-user-config",
		"--sandbox", "workspace-write",
	}
	if spec.Model != "" {
		args = append(args, "--model", spec.Model)
	}
	return args
}

func buildPrompt(spec RunSpec) string {
	if spec.SkillInstalled {
		// Forced mode: explicitly direct the agent to use the installed skill so we
		// measure content quality, not trigger/activation behavior.
		return fmt.Sprintf(
			"Use the skill at .claude/skills/%s/SKILL.md for this task.\n\n%s",
			spec.SkillSlug, spec.Instruction)
	}
	return spec.Instruction
}

func installSkillIntoWorkspace(workDir, slug, content string) error {
	if strings.TrimSpace(content) == "" {
		return fmt.Errorf("treatment arm requires skill content but none was provided")
	}
	dir := filepath.Join(workDir, ".claude", "skills", slug)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, "SKILL.md"), []byte(content), 0o644)
}

// --- usage parsing (pure functions, unit-tested directly) ---

// ParseAgentUsage dispatches to the provider-specific parser.
func ParseAgentUsage(provider string, stdout []byte) (RunOutcome, error) {
	if provider == "claude" {
		return parseClaudeUsage(stdout)
	}
	return parseCodexUsage(stdout)
}

type claudeResult struct {
	TotalCostUSD *float64 `json:"total_cost_usd"`
	NumTurns     int      `json:"num_turns"`
	DurationMS   int64    `json:"duration_ms"`
	IsError      bool     `json:"is_error"`
	Usage        struct {
		InputTokens              int64 `json:"input_tokens"`
		OutputTokens             int64 `json:"output_tokens"`
		CacheReadInputTokens     int64 `json:"cache_read_input_tokens"`
		CacheCreationInputTokens int64 `json:"cache_creation_input_tokens"`
	} `json:"usage"`
}

func parseClaudeUsage(stdout []byte) (RunOutcome, error) {
	var res claudeResult
	if err := json.Unmarshal(stdout, &res); err != nil {
		// claude -p --output-format json emits a single result object; if a stream
		// slipped through, take the last JSON object on the last non-empty line.
		if obj, ok := lastJSONObject(stdout); ok {
			if err2 := json.Unmarshal(obj, &res); err2 != nil {
				return RunOutcome{}, fmt.Errorf("parse claude usage: %w", err)
			}
		} else {
			return RunOutcome{}, fmt.Errorf("parse claude usage: %w", err)
		}
	}
	return RunOutcome{
		InputTokens:  res.Usage.InputTokens,
		OutputTokens: res.Usage.OutputTokens,
		CacheRead:    res.Usage.CacheReadInputTokens,
		CacheWrite:   res.Usage.CacheCreationInputTokens,
		CostUSD:      res.TotalCostUSD,
		DurationMS:   res.DurationMS,
		Turns:        res.NumTurns,
		AgentError:   res.IsError,
	}, nil
}

type codexTurnCompleted struct {
	Type  string `json:"type"`
	Usage struct {
		InputTokens         int64 `json:"input_tokens"`
		CachedInputTokens   int64 `json:"cached_input_tokens"`
		OutputTokens        int64 `json:"output_tokens"`
		ReasoningOutputTokens int64 `json:"reasoning_output_tokens"`
	} `json:"usage"`
}

func parseCodexUsage(stdout []byte) (RunOutcome, error) {
	// codex exec --json emits JSONL; accumulate turn.completed usage and count
	// turns.
	lines := strings.Split(string(stdout), "\n")
	var out RunOutcome
	found := false
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || !strings.HasPrefix(line, "{") {
			continue
		}
		var evt codexTurnCompleted
		if err := json.Unmarshal([]byte(line), &evt); err != nil {
			continue
		}
		if evt.Type == "turn.completed" {
			found = true
			out.InputTokens += evt.Usage.InputTokens
			out.OutputTokens += evt.Usage.OutputTokens + evt.Usage.ReasoningOutputTokens
			out.CacheRead += evt.Usage.CachedInputTokens
			out.Turns++
		}
	}
	if !found {
		return RunOutcome{}, fmt.Errorf("parse codex usage: no turn.completed events")
	}
	return out, nil
}

func lastJSONObject(data []byte) ([]byte, bool) {
	lines := strings.Split(string(data), "\n")
	for i := len(lines) - 1; i >= 0; i-- {
		line := strings.TrimSpace(lines[i])
		if strings.HasPrefix(line, "{") && strings.HasSuffix(line, "}") {
			return []byte(line), true
		}
	}
	return nil, false
}

File: crates/skillrank-core/src/runner/agent.rs
//! Invokes the user's own `claude`/`codex` binary as a one-shot on the user's own
//! subscription. The verifier is applied by the caller AFTER this returns, so the
//! agent workspace never contains verifier content.

use super::{AgentRunner, RunOutcome, RunSpec};
use serde::Deserialize;
use sha2::{Digest, Sha256};
use std::io::{Read, Write};
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

pub struct CliAgentRunner {
    pub provider: String, // "claude" | "codex"
    pub binary: String,
    pub version: String,
}

impl AgentRunner for CliAgentRunner {
    fn agent_name(&self) -> String {
        if self.provider == "claude" {
            "claude_code".to_string()
        } else {
            self.provider.clone()
        }
    }

    fn agent_version_band(&self) -> String {
        version_band(&self.version)
    }

    fn run_task(&self, spec: &RunSpec) -> Result<RunOutcome, String> {
        if spec.skill_installed {
            install_skill_into_workspace(spec)?;
        }
        let prompt = build_prompt(spec);
        let timeout = Duration::from_secs(if spec.timeout_sec == 0 {
            240
        } else {
            spec.timeout_sec as u64
        });

        let mut child = Command::new(&self.binary)
            .args(self.argv(spec))
            .current_dir(&spec.working_dir)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::null())
            .spawn()
            .map_err(|e| format!("spawn {}: {e}", self.binary))?;

        // Drain stdout on a thread so a full pipe never deadlocks us.
        let mut stdout = child.stdout.take().expect("piped stdout");
        let reader = std::thread::spawn(move || {
            let mut buf = Vec::new();
            let _ = stdout.read_to_end(&mut buf);
            buf
        });
        if let Some(mut stdin) = child.stdin.take() {
            let _ = stdin.write_all(prompt.as_bytes());
            // stdin dropped here -> EOF for the child
        }

        let start = Instant::now();
        let mut timed_out = false;
        let mut success = false;
        loop {
            match child.try_wait() {
                Ok(Some(status)) => {
                    success = status.success();
                    break;
                }
                Ok(None) => {
                    if start.elapsed() > timeout {
                        let _ = child.kill();
                        let _ = child.wait();
                        timed_out = true;
                        break;
                    }
                    std::thread::sleep(Duration::from_millis(50));
                }
                Err(_) => break,
            }
        }
        let elapsed = start.elapsed().as_millis() as i64;
        let stdout_bytes = reader.join().unwrap_or_default();
        let digest = format!("sha256:{:x}", Sha256::digest(&stdout_bytes));

        if timed_out || !success {
            return Ok(RunOutcome {
                duration_ms: elapsed,
                trajectory_digest: digest,
                agent_error: true,
                ..Default::default()
            });
        }
        match parse_agent_usage(&self.provider, &stdout_bytes) {
            Ok(mut outcome) => {
                outcome.duration_ms = elapsed;
                outcome.trajectory_digest = digest;
                Ok(outcome)
            }
            Err(_) => Ok(RunOutcome {
                duration_ms: elapsed,
                trajectory_digest: digest,
                agent_error: true,
                ..Default::default()
            }),
        }
    }
}

impl CliAgentRunner {
    fn argv(&self, spec: &RunSpec) -> Vec<String> {
        if self.provider == "claude" {
            let mut args = vec![
                "-p".to_string(),
                "--output-format".to_string(),
                "json".to_string(),
                "--dangerously-skip-permissions".to_string(),
            ];
            if !spec.model.is_empty() {
                args.push("--model".into());
                args.push(spec.model.clone());
            }
            args
        } else {
            let mut args = vec![
                "exec".to_string(),
                "--json".to_string(),
                "--skip-git-repo-check".to_string(),
                "--ignore-user-config".to_string(),
                "--sandbox".to_string(),
                "workspace-write".to_string(),
            ];
            if !spec.model.is_empty() {
                args.push("--model".into());
                args.push(spec.model.clone());
            }
            args
        }
    }
}

/// Reduce a semver-ish version string to "major.minor".
pub fn version_band(version: &str) -> String {
    let v = version.trim().trim_start_matches('v');
    let parts: Vec<&str> = v.splitn(3, '.').collect();
    if parts.len() >= 2 {
        format!("{}.{}", parts[0], parts[1])
    } else {
        v.to_string()
    }
}

fn build_prompt(spec: &RunSpec) -> String {
    if spec.skill_installed {
        // Forced mode: explicitly direct the agent to use the installed skill so we
        // measure content quality, not trigger/activation behavior.
        format!(
            "Use the skill at .claude/skills/{}/SKILL.md for this task.\n\n{}",
            spec.skill_slug, spec.instruction
        )
    } else {
        spec.instruction.clone()
    }
}

fn install_skill_into_workspace(spec: &RunSpec) -> Result<(), String> {
    if spec.skill_content.trim().is_empty() {
        return Err("treatment arm requires skill content but none was provided".into());
    }
    let dir = spec
        .working_dir
        .join(".claude")
        .join("skills")
        .join(&spec.skill_slug);
    std::fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    std::fs::write(dir.join("SKILL.md"), &spec.skill_content).map_err(|e| e.to_string())
}

// ---- usage parsing (pure, unit-tested directly) ----

/// Dispatch to the provider-specific parser.
pub fn parse_agent_usage(provider: &str, stdout: &[u8]) -> Result<RunOutcome, String> {
    if provider == "claude" {
        parse_claude_usage(stdout)
    } else {
        parse_codex_usage(stdout)
    }
}

#[derive(Deserialize, Default)]
struct ClaudeUsage {
    #[serde(default)]
    input_tokens: i64,
    #[serde(default)]
    output_tokens: i64,
    #[serde(default)]
    cache_read_input_tokens: i64,
    #[serde(default)]
    cache_creation_input_tokens: i64,
}

#[derive(Deserialize, Default)]
struct ClaudeResult {
    #[serde(default)]
    total_cost_usd: Option<f64>,
    #[serde(default)]
    num_turns: i64,
    #[serde(default)]
    duration_ms: i64,
    #[serde(default)]
    is_error: bool,
    #[serde(default)]
    usage: ClaudeUsage,
}

fn parse_claude_usage(stdout: &[u8]) -> Result<RunOutcome, String> {
    let res: ClaudeResult = match serde_json::from_slice(stdout) {
        Ok(r) => r,
        Err(e) => {
            // If a stream slipped through, take the last JSON object line.
            match last_json_object(stdout) {
                Some(obj) => {
                    serde_json::from_slice(&obj).map_err(|_| format!("parse claude usage: {e}"))?
                }
                None => return Err(format!("parse claude usage: {e}")),
            }
        }
    };
    Ok(RunOutcome {
        input_tokens: res.usage.input_tokens,
        output_tokens: res.usage.output_tokens,
        cache_read: res.usage.cache_read_input_tokens,
        cache_write: res.usage.cache_creation_input_tokens,
        cost_usd: res.total_cost_usd,
        duration_ms: res.duration_ms,
        turns: res.num_turns,
        agent_error: res.is_error,
        trajectory_digest: String::new(),
    })
}

#[derive(Deserialize, Default)]
struct CodexUsage {
    #[serde(default)]
    input_tokens: i64,
    #[serde(default)]
    cached_input_tokens: i64,
    #[serde(default)]
    output_tokens: i64,
    #[serde(default)]
    reasoning_output_tokens: i64,
}

#[derive(Deserialize, Default)]
struct CodexEvent {
    #[serde(default, rename = "type")]
    event_type: String,
    #[serde(default)]
    usage: CodexUsage,
}

fn parse_codex_usage(stdout: &[u8]) -> Result<RunOutcome, String> {
    let text = String::from_utf8_lossy(stdout);
    let mut out = RunOutcome::default();
    let mut found = false;
    for line in text.split('\n') {
        let line = line.trim();
        if line.is_empty() || !line.starts_with('{') {
            continue;
        }
        let Ok(evt): Result<CodexEvent, _> = serde_json::from_str(line) else {
            continue;
        };
        if evt.event_type == "turn.completed" {
            found = true;
            out.input_tokens += evt.usage.input_tokens;
            out.output_tokens += evt.usage.output_tokens + evt.usage.reasoning_output_tokens;
            out.cache_read += evt.usage.cached_input_tokens;
            out.turns += 1;
        }
    }
    if !found {
        return Err("parse codex usage: no turn.completed events".into());
    }
    Ok(out)
}

fn last_json_object(data: &[u8]) -> Option<Vec<u8>> {
    let text = String::from_utf8_lossy(data);
    for line in text.split('\n').rev() {
        let line = line.trim();
        if line.starts_with('{') && line.ends_with('}') {
            return Some(line.as_bytes().to_vec());
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_claude_usage_works() {
        let stdout = br#"{"type":"result","total_cost_usd":0.1234,"num_turns":5,"duration_ms":8200,"is_error":false,"usage":{"input_tokens":1200,"output_tokens":800,"cache_read_input_tokens":5000,"cache_creation_input_tokens":300}}"#;
        let out = parse_claude_usage(stdout).unwrap();
        assert_eq!(out.input_tokens, 1200);
        assert_eq!(out.output_tokens, 800);
        assert_eq!(out.cache_read, 5000);
        assert_eq!(out.cache_write, 300);
        assert_eq!(out.cost_usd, Some(0.1234));
        assert_eq!(out.turns, 5);
    }

    #[test]
    fn parse_codex_usage_accumulates() {
        let stdout = b"{\"type\":\"thread.started\"}\n{\"type\":\"turn.completed\",\"usage\":{\"input_tokens\":500,\"cached_input_tokens\":100,\"output_tokens\":200,\"reasoning_output_tokens\":50}}\n{\"type\":\"turn.completed\",\"usage\":{\"input_tokens\":300,\"cached_input_tokens\":80,\"output_tokens\":150,\"reasoning_output_tokens\":25}}";
        let out = parse_codex_usage(stdout).unwrap();
        assert_eq!(out.input_tokens, 800);
        assert_eq!(out.output_tokens, 425);
        assert_eq!(out.cache_read, 180);
        assert_eq!(out.turns, 2);
    }

    #[test]
    fn parse_codex_usage_no_turns_errors() {
        assert!(parse_codex_usage(b"{\"type\":\"thread.started\"}").is_err());
    }

    #[test]
    fn version_band_works() {
        assert_eq!(version_band("2.1.174"), "2.1");
        assert_eq!(version_band("v2.1.176"), "2.1");
        assert_eq!(version_band("1.0"), "1.0");
        assert_eq!(version_band("3"), "3");
    }

    #[test]
    fn build_prompt_forced_vs_control() {
        let control = build_prompt(&RunSpec {
            working_dir: ".".into(),
            instruction: "fix".into(),
            model: String::new(),
            skill_installed: false,
            skill_content: String::new(),
            skill_slug: String::new(),
            timeout_sec: 0,
        });
        assert_eq!(control, "fix");
        let treatment = build_prompt(&RunSpec {
            working_dir: ".".into(),
            instruction: "fix".into(),
            model: String::new(),
            skill_installed: true,
            skill_content: String::new(),
            skill_slug: "owner/skill".into(),
            timeout_sec: 0,
        });
        assert_ne!(treatment, control);
        assert!(treatment.contains(".claude/skills/owner/skill/SKILL.md"));
    }
}
